pkg/globalnet/controllers/ingress_endpoints_controller.go

```diff
@@ -0,0 +1,142 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+*/
+
+package controllers
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+	"github.com/submariner-io/admiral/pkg/federate"
+	"github.com/submariner-io/admiral/pkg/stringset"
+	"github.com/submariner-io/admiral/pkg/syncer"
+	"github.com/submariner-io/admiral/pkg/util"
+	submarinerv1 "github.com/submariner-io/submariner/pkg/apis/submariner.io/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+)
+
+func startIngressEndpointsController(svc *corev1.Service, config *syncer.ResourceSyncerConfig) (*ingressEndpointsController, error) {
+	var err error
+
+	_, gvr, err := util.ToUnstructuredResource(&submarinerv1.GlobalIngressIP{}, config.RestMapper)
+	if err != nil {
+		return nil, errors.Wrap(err, "error converting resource")
+	}
+
+	controller := &ingressEndpointsController{
+		baseSyncerController: newBaseSyncerController(),
+		svcName:              svc.Name,
+		namespace:            svc.Namespace,
+		ingressIPMap:         stringset.NewSynchronized(),
+	}
+
+	fieldSelector := fields.Set(map[string]string{"metadata.name": svc.Name}).AsSelector().String()
+
+	controller.resourceSyncer, err = syncer.NewResourceSyncer(&syncer.ResourceSyncerConfig{
+		Name:                "Ingress Endpoints syncer",
+		ResourceType:        &corev1.Endpoints{},
```

Comment:
Can you please explain the impact on this code when the endpoints are deprecated/removed and replaced with EndpointSlices?

Reply:
If `Endpoints` are deprecated, we would need to re-implement this to use `EndpointSlices`. However, I'm not sure if there are fixed plan to deprecate it.

For this particular headless service without selector case, I think that we should use `Endpoints`, because `"The control plane automatically creates EndpointSlices for any Kubernetes Service that has a selector specified."` as described [here](https://kubernetes.io/docs/concepts/services-networking/endpoint-slices/#endpointslice-resource). As I tested as below, `EndpointSlices` doesn't seem to be created automatically, for manually created `Endpoints`. 

<details>

```bash
kubectl get endpoints,endpointslices
NAME                   ENDPOINTS         AGE
endpoints/kubernetes   172.18.0.2:6443   2m39s

NAME                                        ADDRESSTYPE   PORTS   ENDPOINTS    AGE
endpointslice.discovery.k8s.io/kubernetes   IPv4          6443    172.18.0.2   2m39s

cat << EOF | kubectl apply -f -
apiVersion: v1
kind: Endpoints
metadata:
  name: test-vm
subsets:
  - addresses:
      - ip: 192.168.4.2
    ports:
      - port: 80
EOF

kubectl get endpoints,endpointslices
NAME                   ENDPOINTS         AGE
endpoints/kubernetes   172.18.0.2:6443   3m48s
endpoints/test-vm      192.168.4.2:80    4s

NAME                                        ADDRESSTYPE   PORTS   ENDPOINTS    AGE
endpointslice.discovery.k8s.io/kubernetes   IPv4          6443    172.18.0.2   3m48s
```
</details>

Reply:
Ok, thanks @mkimuram